File: Morse/project/morse/IOHelper.m
//
//  IOHelper.m
//  morse
//
//  Created by Gabriel Bremond on 15/12/2013.
//  Copyright (c) 2013 Gabriel Bremond. All rights reserved.
//

#import "IOHelper.h"

@implementation IOHelper

#pragma mark -
#pragma mark Public methods

// Search for data in given arguments.
+ (MorseOptions)lookForData:(NSString **)data fromArgs:(const char *[])argv
{
    MorseOptions opts = MorseOptionMode;
    int lastOption = 0;
    
    @try
    {
        NSString *arg;
        for (int i=0; i < sizeof(&argv); i++)
        {
            if (argv[i] == nil)
            {
                continue;
            }
            
            // Grab argument.
            arg = [NSString stringWithUTF8String:argv[i]];
            
            // Look for help options.
            if (([arg isEqualToString:@"--usage"] == YES) || ([arg isEqualToString:@"-help"] == YES) || ([arg isEqualToString:@"-h"] == YES))
            {
                [IOHelper printHelp];
            }
            // Look for version.
            else if ([arg isEqualToString:@"--version"] == YES)
            {
                [IOHelper printVersion];
            }
            // Tread other options.
            else if ([arg characterAtIndex:0] == '-')
            {
                for (int j=0; j<arg.length; j++)
                {
                    // Check for deciphe.
                    if ([arg characterAtIndex:j] == 'd')
                    {
                        opts ^= MorseOptionMode;
                        lastOption = i;
                    }
                    // Check verbose mode.
                    if ([arg characterAtIndex:j] == 'v')
                    {
                        opts |= MorseOptionVerbose;
                        lastOption = i;
                    }
                    // Check copy option.
                    if ([arg characterAtIndex:j] == 'c')
                    {
                        opts |= MorseOptionPastboard;
                        lastOption = i;
                    }
                }
            }
        }
    }
    @catch (NSException *exception)
    {
    }

    // Last option cannot be in an offset upper than the arguments size.
    if (sizeof(&argv) <= lastOption)
    {
        [IOHelper exitOnError:@"Empty data"];
    }
    
    // Convert data into an NSString.
    @try
    {
        *data = [NSString stringWithUTF8String:argv[lastOption + 1]];
        
        // Check data format.
        if (*data==NULL || [*data length]==0)
        {
            [[NSException exceptionWithName:@"Data error" reason:@"Failed to retrieve data." userInfo:nil] raise];
        }
    }
    @catch (NSException *exception)
    {
        [IOHelper exitOnError:@"Empty data"];
    }
    
    return opts;
}

// Exit the programdisplaying the error message and invite user to use the help option.
+ (void)exitOnError:(NSString *)message
{
    printf("%s\n\r",[message cStringUsingEncoding:NSUTF8StringEncoding]);
    printf("Use: 'morse -h' for help.\n\r");
    exit(-1);
}


#pragma mark -
#pragma mark Private methods

+ (void)printHelp
{
    printf("\n\r");
    printf("Morse ©photograve 2013.\n\r");
    printf("This is a MacOSX command line.\n\r");
    printf("\n\r");
    printf("Use input parameters: \n\r");
    printf("    -d    to decode a Morse string (optional).\n\r");
    printf("    -v    Verbose mode (optional).\n\r");
    printf("    -c    Copy result into the general pastboard (optional).");
    printf("\n\r");
    printf("morse [-dvc] \"My string here\"\n\r");
    printf("\n\r");
    printf("--version          Get the program version.");
    printf("\n\r");
    printf("--usage | -help    To get this help.\n\r");
    printf("\n\r");
    exit(-1);
}

+ (void)printVersion
{
    printf("\n\r");
    printf("Morse CLI: \n\r");
    printf("Version 1.2\n\r");
    printf("\n\r");
    exit(-1);
}

@end

File: Morse/project-swift/morse/IOHelper.swift
//
//  IOHelper.swift
//  morse
//
//  Created by Gabriel Bremond on 05/06/14.
//  Copyright (c) 2014 photograve. All rights reserved.
//

import Foundation

enum MorseOptions {
    
    case mode
    case verbose
    case pastboard
}

class IOHelper {
    
    class func lookForData() -> (String, [MorseOptions]) {
        
        var opts = [MorseOptions]()
        var lastOption = 0
        
        var arg: String
        var number: Int = 0
        
        for i in 0...(CommandLine.arguments.count - 1)
        {
            let index = Int(i)
            arg = CommandLine.arguments[index]
 
            if arg == "--usage" || arg == "-help" || arg == "-h"
            {
                IOHelper.printHelp()
            }
            else if arg == "--version"
            {
                IOHelper.printVersion()
            }
            else if arg.hasPrefix("-")
            {
                for char: Character in arg.characters
                {
                    if String(char) == "d"
                    {
                        opts.append(MorseOptions.mode)
                        lastOption = Int(i)
                    }
                    else if String(char) == "v"
                    {
                        opts.append(MorseOptions.verbose)
                        lastOption = Int(i)
                    }
                    else if String(char) == "c"
                    {
                        opts.append(MorseOptions.pastboard)
                        lastOption = Int(i)
                    }
                }
            }
            
            number = Int(i)
        }

        // Check ranges.
        let lastIndex = Int(lastOption + 1)
        if number < lastIndex
        {
            IOHelper.exitOnError("Empty data")
        }
        
        // Look for content.
        let content = CommandLine.arguments[lastIndex]
        if content.isEmpty
        {
            IOHelper.exitOnError("Empty data")
        }

        // Last option cannot be empty.
        arg = String.init(validatingUTF8:content)!
        if arg.isEmpty
        {
            IOHelper.exitOnError("Empty data")
        }
        
        return (arg, opts)
    }
    
    class func exitOnError(_ message: String)
    {
        print("")
        print("\(message)")
        print("Use: 'morse -h' for help.")
        print("")

        exit(-1)
    }
    
    class func printHelp()
    {
        print("")
        print("Morse ©photograve 2016.")
        print("http://photograve.fr")
        print("This is a macOS command line.")
        print("")
        print("Use input parameters:")
        print("    -d    to decode a Morse string (optional).")
        print("    -v    Verbose mode (optional).")
        print("    -c    Copy result into the general pastboard (optional).")
        print("")
        print("morse [-dvc] \"My string here\"")
        print("")
        print("--version          Get the program version.")
        print("")
        print("--usage | -help    To get this help.")
        print("")
        
        exit(0)
    }
    
    class func printVersion()
    {
        print("")
        print("Morse CLI:")
        print("Version 1.5")
        print("")
        
        exit(0)
    }
}
